scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala

```diff
@@ -280,47 +328,113 @@ class Randomizer(private[scalatest] val seed: Long) { thisRandomizer =>
+    // exponent therefore is 254, which in hex is 0xfe (which represents an exponent of +127).
+    // Thus by chosing the exponent Int between 0 and 0xfe, we can't get a NaN or an infinity.
+    val (s, rs) = chooseInt(0, 1)        // The sign bit (1 bit)
+    val (e, re) = chooseInt(0, 0xfe)     // The exponent (8 bits)
+    val (m, rm) = chooseInt(0, 0x7fffff) // The mantissa (23 bits)
+    val finite = java.lang.Float.intBitsToFloat((s << 31) | (e << 23) | m)
+    (FiniteFloat.ensuringValid(finite), rm)
+  }
+
+  private def nextFiniteFloatValue: (Float, Randomizer) = {
+    val (ff, r) = nextFiniteFloat
+    (ff.value, r)
+  }
+
+  // An extended real float includes all the numeric ones plus positive and negative infinity
+  // In other words, the whole space other than the NaNs. This may be a missing abstraction in anyvals.
+  // Let's think about that.
+  private def nextExtRealFloatValue: (Float, Randomizer) = {
+
+    // The space of possible non-NaN Floats
+    // 2 to the power of 32, which is covered by Int.MinValue to Int.MaxValue
+    // - (2 to the power of 23 (23 is the number of mantissa bits))
+    // + (2 for the two infinities)
+
+    // scala> Int.MaxValue - math.pow(2, 23).toInt + 2
+    // res13: Int = 2139095041
+    val (x, r) = chooseInt(Int.MinValue, 2139095041) 
+
+    // Pick two lucky numbers to play the lotto with:
+    if (x == 777)
+      (Float.NegativeInfinity, r)
+    else if (x == 888)
+      (Float.PositiveInfinity, r)
+    else r.nextFiniteFloatValue
   }
 
   /**
     * Get a random non-infinite Double.
     *
     * This can return either a positive or negative value, or zero, but guards against
-    * returning either [[Double.PositiveInfinity]] or [[Double.NegativeInfinity]].
+    * returning either [[Double.PositiveInfinity]], [[Double.NegativeInfinity]], or [[Double.NaN]].
     *
     * @return A random finite Double, and the next Randomizer to use.
     */
   def nextFiniteDouble: (FiniteDouble, Randomizer) = {
-    val (n, r) = nextDouble // TODO: Study nextFloat and nextDouble to see if it produces NaNs or Infinities.
-    val finite =            // See if it produces non-normal (less than max precision) values
-      n match {
-        case Double.PositiveInfinity => Double.MaxValue
-        case Double.NegativeInfinity => Double.MaxValue
-        case _ => n
-      }
-    (FiniteDouble.ensuringValid(finite), r)
+    // The exponent portion of a Double occupies 11 bits. It can be 0 (which represents an exponent of -1023)
+    // to 2047 (which is a reserved value for NaN values and +/- infinity). The highest regular (non-reserved)
+    // exponent therefore is 2046, which in hex is 0x7fe (which represents an exponent of +1023).
+    // Thus by chosing the exponent Int between 0 and 0x7fe, we can't get a NaN or an infinity.
+    val (s, rs) = chooseLong(0L, 1L)                  // The sign bit (1 bit)
```

Comment:
Is `chooseLong()` the right tool for this job?  Seems like an awful lot of cycles to choose between 0 and 1.  We might want a simple `nextBit()` for this.

Reply:
I'll look at next(1).